diff -Naur ../NetworkManager-1.12.2-unpatched/src/dhcp/nm-dhcp-dhcpcd.c ./src/dhcp/nm-dhcp-dhcpcd.c
--- ../NetworkManager-1.12.2-unpatched/src/dhcp/nm-dhcp-dhcpcd.c	2019-01-08 15:41:57.052427624 +0100
+++ ./src/dhcp/nm-dhcp-dhcpcd.c	2019-01-08 15:42:40.244359485 +0100
@@ -118,10 +118,10 @@
 
 	g_ptr_array_add (argv, (gpointer) "-K");    /* Disable built-in carrier detection */
 
-	g_ptr_array_add (argv, (gpointer) "-L");    /* Disable built-in IPv4LL */
+	/* g_ptr_array_add (argv, (gpointer) "-L");    /\* Disable built-in IPv4LL *\/ */
 
 	/* --noarp. Don't request or claim the address by ARP; this also disables IPv4LL. */
-	g_ptr_array_add (argv, (gpointer) "-A");
+	/* g_ptr_array_add (argv, (gpointer) "-A"); */
 
 	g_ptr_array_add (argv, (gpointer) "-G");    /* Let NM handle routing */
 
